make/file_targets.cmake

```diff
@@ -0,0 +1,265 @@
+    message(
+      FATAL_ERROR
+        "File ${src_file} is not a valid verilog target, missing LOCATION."
+    )
+  endif()
+
+  list(APPEND ${var} ${SRC_TARGET})
+  list(APPEND ${var} ${SRC_LOCATION})
+  set(${var} "${${var}}" PARENT_SCOPE)
+endfunction()
+
+function(APPEND_FILE_INCLUDES var src_file)
+  # Appends to list var in PARENT_SCOPE both includes listed for file target for
+  # given src_file.
+  get_file_target(SRC_TARGET ${src_file})
+  get_target_property(SRC_INCLUDES ${SRC_TARGET} INCLUDES)
+  if("${SRC_INCLUDES}" STREQUAL "NOT_FOUND")
+    message(
+      FATAL_ERROR
+        "File ${SRC} is not a valid verilog target, missing INCLUDES list."
+    )
+  endif()
+
+  list(APPEND ${var} ${SRC_INCLUDES})
+  set(${var} "${${var}}" PARENT_SCOPE)
+endfunction()
+
+function(GET_VERILOG_INCLUDES var file)
+  # Appends to list var in PARENT_SCOPE all verilog source dependency based on
+  # scanning input at configure time.
+  #
+  # Note this function cannot be used at generation time because execute_process
+  # is called during configure step and generated files don't exist yet.
+  execute_process(
+    COMMAND grep "^`include" "${file}"
+    COMMAND
+      cut -d \" -f 2
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    OUTPUT_VARIABLE INCLUDES
+  )
+
+  string(
+    REPLACE
+      "\n"
+      ";"
+      INCLUDES_LIST
+      "${INCLUDES}"
+  )
+  foreach(INCLUDE ${INCLUDES_LIST})
+    string(STRIP ${INCLUDE} INCLUDE)
+    if(NOT "${INCLUDE}" STREQUAL "")
+      get_filename_component(
+        ABS_PATH_TO_INCLUDE
+        ${CMAKE_CURRENT_SOURCE_DIR}/${INCLUDE}
+        ABSOLUTE
+        BASE_DIR
+        ${symbiflow-arch-defs_SOURCE_DIR}
+      )
+      list(APPEND ${var} ${ABS_PATH_TO_INCLUDE})
+    endif()
+  endforeach()
+  set(${var} "${${var}}" PARENT_SCOPE)
+endfunction()
+
+function(GET_XML_INCLUDES var file)
+  # Appends to list var in PARENT_SCOPE all xml source dependency based on
+  # scanning input at configure time.
+  #
+  # Note this function cannot be used at generation time because execute_process
+  # is called during configure step and generated files don't exist yet.
+  get_target_property_required(PYTHON3 env PYTHON3)
+  execute_process(
+    COMMAND
+      ${PYTHON3} ${symbiflow-arch-defs_SOURCE_DIR}/utils/deps_xml.py
+      --file_per_line ${CMAKE_CURRENT_SOURCE_DIR}/${file}
+    WORKING_DIRECTORY ${symbiflow-arch-defs_SOURCE_DIR}
+    OUTPUT_VARIABLE INCLUDES
+  )
+
+  string(
+    REPLACE
+      "\n"
+      ";"
+      INCLUDES_LIST
+      "${INCLUDES}"
+  )
+  foreach(INCLUDE ${INCLUDES_LIST})
+    string(STRIP ${INCLUDE} INCLUDE)
+    if(NOT "${INCLUDE}" STREQUAL "")
+      get_filename_component(
+        ABS_PATH_TO_INCLUDE
+        ${INCLUDE}
+        ABSOLUTE
+        BASE_DIR
+        ${symbiflow-arch-defs_SOURCE_DIR}
+      )
+      list(APPEND ${var} ${ABS_PATH_TO_INCLUDE})
+    endif()
+  endforeach()
+  set(${var} "${${var}}" PARENT_SCOPE)
+endfunction()
+
+function(MAKE_FILE_TARGET)
```

Comment:
Maybe this should be `CREATE_FILE_TARGET` -- I keep getting it confused is `makefile target` or something?